modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java

```diff
@@ -0,0 +1,365 @@
+            LOGGER.debug("database update [{}] already in progress, skipping...", databaseName);
+            return;
+        }
+
+        final Path databaseTmpFile = Files.createFile(geoipTmpDirectory.resolve(databaseName + ".tmp"));
+        LOGGER.info("downloading geoip database [{}] to [{}]", databaseName, databaseTmpGzFile);
+        retrieveDatabase(
+            databaseName,
+            recordedMd5,
+            metadata,
+            bytes -> Files.write(databaseTmpGzFile, bytes, StandardOpenOption.APPEND),
+            () -> {
+                LOGGER.debug("decompressing [{}]", databaseTmpGzFile.getFileName());
+                decompress(databaseTmpGzFile, databaseTmpFile);
+
+                Path databaseFile = geoipTmpDirectory.resolve(databaseName);
+                LOGGER.debug("moving database from [{}] to [{}]", databaseTmpFile, databaseFile);
+                Files.move(databaseTmpFile, databaseFile, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
+                updateDatabase(databaseName, recordedMd5, databaseFile);
+                Files.delete(databaseTmpGzFile);
+            },
+            failure -> {
+                LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to download database [{}]", databaseName), failure);
+                try {
+                    Files.delete(databaseTmpFile);
+                    Files.delete(databaseTmpGzFile);
+                } catch (IOException ioe) {
+                    ioe.addSuppressed(failure);
+                    LOGGER.error("Unable to delete tmp database file after failure", ioe);
+                }
+            });
+    }
+
+    void updateDatabase(String databaseFileName, String recordedMd5, Path file) {
+        try {
+            LOGGER.info("database file changed [{}], reload database...", file);
+            DatabaseReaderLazyLoader loader = new DatabaseReaderLazyLoader(cache, file);
+            DatabaseReference existing = databases.put(databaseFileName, new DatabaseReference(recordedMd5, file, loader));
+            if (existing != null) {
+                existing.close();
+            }
+        } catch (Exception e) {
+            LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to update database [{}]", databaseFileName), e);
+        }
+    }
+
+    void removeStaleEntries(Collection<String> staleEntries) {
+        try {
+            List<DatabaseReference> references = new ArrayList<>();
+            for (String staleEntry : staleEntries) {
+                LOGGER.info("database [{}] no longer exists, cleaning up...", staleEntry);
+                DatabaseReference existing = databases.remove(staleEntry);
+                assert existing != null;
+                references.add(existing);
+            }
+
+            for (DatabaseReference reference : references) {
+                reference.close();
+                Files.delete(reference.databaseFile);
```

Comment:
will it mess something if we delete the file during concurrent lookup? `refence.close()` may not actually close anything if there are any  ongoing lookups

Reply:
I think *unix like systems handle this gracefully (deleting a file in use, only after its file descriptor has been closed). So I think the current will not mess anything up, but windows not so sure. In anyway perhaps it makes more sense to move the actual deletion of the db file elsewhere (maybe to `DatabaseReaderLazyLoader#doClose()` are at least initiated from there).